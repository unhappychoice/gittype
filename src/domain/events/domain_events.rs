use super::Event;
use std::any::Any;

// Unified domain event enum for pattern matching
#[derive(Debug, Clone)]
pub enum DomainEvent {
    KeyPressed { key: char, position: usize },
    StageStarted { start_time: std::time::Instant },
    StagePaused,
    StageResumed,
    StageFinalized,
    StageSkipped,
    ChallengeLoaded { text: String, source_path: String },
}

impl Event for DomainEvent {
    fn as_any(&self) -> &dyn Any {
        self
    }
}
