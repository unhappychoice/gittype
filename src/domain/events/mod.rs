use std::any::{Any, TypeId};
use std::collections::HashMap;
use std::sync::{Arc, RwLock};

pub mod domain_events;

pub trait Event: Send + Sync + 'static {
    fn as_any(&self) -> &dyn Any;
}

pub trait EventHandler<E: Event>: Send + Sync {
    fn handle(&self, event: &E);
}

type BoxedHandler = Arc<dyn Fn(&dyn Event) + Send + Sync>;

pub struct EventBus {
    subscribers: Arc<RwLock<HashMap<TypeId, Vec<BoxedHandler>>>>,
}

impl EventBus {
    pub fn new() -> Self {
        Self {
            subscribers: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    pub fn publish<E: Event>(&self, event: E) {
        let type_id = TypeId::of::<E>();

        // Clone Arc<handlers> while holding the lock, then release it before calling them
        let handlers: Vec<BoxedHandler> = {
            let subscribers = self.subscribers.read().unwrap();
            let h = subscribers.get(&type_id)
                .map(|h| h.clone())
                .unwrap_or_default();
            h
        }; // Lock is released here

        // Call handlers without holding the lock
        for handler in handlers.iter() {
            handler(&event);
        }
    }

    pub fn subscribe<E: Event, F>(&self, handler: F)
    where
        F: Fn(&E) + Send + Sync + 'static,
    {
        let type_id = TypeId::of::<E>();
        let mut subscribers = self.subscribers.write().unwrap();

        let boxed_handler: BoxedHandler = Arc::new(move |event: &dyn Event| {
            if let Some(concrete_event) = event.as_any().downcast_ref::<E>() {
                handler(concrete_event);
            }
        });

        subscribers
            .entry(type_id)
            .or_insert_with(Vec::new)
            .push(boxed_handler);
    }
}

impl Default for EventBus {
    fn default() -> Self {
        Self::new()
    }
}

impl Clone for EventBus {
    fn clone(&self) -> Self {
        Self {
            subscribers: Arc::clone(&self.subscribers),
        }
    }
}
