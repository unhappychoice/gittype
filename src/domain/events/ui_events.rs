use super::Event;
use std::any::Any;

#[derive(Debug, Clone)]
pub struct RenderRequested {
    pub screen_name: String,
}

impl Event for RenderRequested {
    fn as_any(&self) -> &dyn Any {
        self
    }
}

#[derive(Debug, Clone)]
pub struct ScreenChanged {
    pub from: String,
    pub to: String,
}

impl Event for ScreenChanged {
    fn as_any(&self) -> &dyn Any {
        self
    }
}

#[derive(Debug, Clone)]
pub struct UserInputReceived {
    pub input: String,
}

impl Event for UserInputReceived {
    fn as_any(&self) -> &dyn Any {
        self
    }
}

#[derive(Debug, Clone)]
pub struct ConfigUpdated {
    pub key: String,
    pub value: String,
}

impl Event for ConfigUpdated {
    fn as_any(&self) -> &dyn Any {
        self
    }
}
