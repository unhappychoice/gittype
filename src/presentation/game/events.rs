use std::any::Any;

// Re-export ScreenTransition as NavigateTo event
pub use super::models::screen::ScreenTransition as NavigateTo;

/// Event emitted when user requests to exit the application (Ctrl+C)
#[derive(Clone, Debug)]
pub struct ExitRequested;

impl crate::domain::events::Event for ExitRequested {
    fn as_any(&self) -> &dyn Any {
        self
    }
}
