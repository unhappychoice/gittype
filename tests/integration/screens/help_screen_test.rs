use crossterm::event::{KeyCode, KeyEvent, KeyModifiers};
use gittype::domain::events::EventBus;
use gittype::presentation::game::events::NavigateTo;
use gittype::presentation::tui::screens::help_screen::HelpScreen;

screen_snapshot_test!(
    test_help_screen_snapshot_cli,
    HelpScreen,
    HelpScreen::new(EventBus::new())
);

screen_snapshot_test!(
    test_help_screen_snapshot_scoring,
    HelpScreen,
    HelpScreen::new(EventBus::new()),
    keys = [KeyEvent::new(KeyCode::Right, KeyModifiers::empty())]
);

screen_snapshot_test!(
    test_help_screen_snapshot_ranks,
    HelpScreen,
    HelpScreen::new(EventBus::new()),
    keys = [
        KeyEvent::new(KeyCode::Right, KeyModifiers::empty()),
        KeyEvent::new(KeyCode::Right, KeyModifiers::empty())
    ]
);

screen_snapshot_test!(
    test_help_screen_snapshot_game_help,
    HelpScreen,
    HelpScreen::new(EventBus::new()),
    keys = [
        KeyEvent::new(KeyCode::Right, KeyModifiers::empty()),
        KeyEvent::new(KeyCode::Right, KeyModifiers::empty()),
        KeyEvent::new(KeyCode::Right, KeyModifiers::empty())
    ]
);

screen_snapshot_test!(
    test_help_screen_snapshot_community,
    HelpScreen,
    HelpScreen::new(EventBus::new()),
    keys = [
        KeyEvent::new(KeyCode::Right, KeyModifiers::empty()),
        KeyEvent::new(KeyCode::Right, KeyModifiers::empty()),
        KeyEvent::new(KeyCode::Right, KeyModifiers::empty()),
        KeyEvent::new(KeyCode::Right, KeyModifiers::empty())
    ]
);

// Event-producing key tests
use crate::integration::screens::helpers::EmptyMockProvider;

screen_key_event_test!(
    test_help_screen_esc_navigates_back,
    HelpScreen,
    NavigateTo,
    KeyCode::Esc,
    KeyModifiers::empty(),
    EmptyMockProvider
);

screen_key_event_test!(
    test_help_screen_ctrl_c_exits,
    HelpScreen,
    NavigateTo,
    KeyCode::Char('c'),
    KeyModifiers::CONTROL,
    EmptyMockProvider
);

// Non-event key tests
screen_key_tests!(
    HelpScreen,
    EmptyMockProvider,
    [
        (
            test_help_screen_left_switches_tab,
            KeyCode::Left,
            KeyModifiers::empty()
        ),
        (
            test_help_screen_h_switches_tab,
            KeyCode::Char('h'),
            KeyModifiers::empty()
        ),
        (
            test_help_screen_right_switches_tab,
            KeyCode::Right,
            KeyModifiers::empty()
        ),
        (
            test_help_screen_l_switches_tab,
            KeyCode::Char('l'),
            KeyModifiers::empty()
        ),
        (
            test_help_screen_up_scrolls,
            KeyCode::Up,
            KeyModifiers::empty()
        ),
        (
            test_help_screen_k_scrolls,
            KeyCode::Char('k'),
            KeyModifiers::empty()
        ),
        (
            test_help_screen_down_scrolls,
            KeyCode::Down,
            KeyModifiers::empty()
        ),
        (
            test_help_screen_j_scrolls,
            KeyCode::Char('j'),
            KeyModifiers::empty()
        ),
        (
            test_help_screen_g_opens_github,
            KeyCode::Char('g'),
            KeyModifiers::empty()
        ),
    ]
);

// Basic methods test
screen_basic_methods_test!(
    test_help_screen_basic_methods,
    HelpScreen,
    HelpScreen::new(EventBus::new()),
    gittype::presentation::tui::ScreenType::Help,
    false
);
