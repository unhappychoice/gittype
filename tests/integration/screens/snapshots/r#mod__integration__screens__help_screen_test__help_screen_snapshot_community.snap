---
source: tests/integration/screens/help_screen_test.rs
expression: output
---
┌Help──────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ CLI Usage │ Scoring System │ Rank System │ Game Help │ Community │ About & Credits │ Third-Party Licenses            │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│  Social Media:                                                                                                       ▲
│                                                                                                                      █
│  Share your progress with #gittype                                                                                   █
│  https://x.com/search?q=%23gittype                                                                                   █
│                                                                                                                      █
│  Join the Community!                                                                                                 █
│                                                                                                                      █
│  GitHub Repository:                                                                                                  █
│  https://github.com/unhappychoice/gittype                                                                            █
│                                                                                                                      █
│  ⭐  Star the repository if you enjoy GitType!                                                                        █
│                                                                                                                      █
│  Contributing:                                                                                                       █
│                                                                                                                      █
│  • Report bugs and suggest features via GitHub Issues                                                                █
│  • Submit pull requests for improvements                                                                             █
│  • Add support for new programming languages                                                                         █
│  • Improve code extraction algorithms                                                                                █
│  • Enhance UI/UX design                                                                                              █
│                                                                                                                      █
│  Bug Reporting:                                                                                                      █
│                                                                                                                      █
│  When reporting bugs, please include:                                                                                █
│  • Operating system and terminal details                                                                             █
│  • Steps to reproduce the issue                                                                                      █
│  • Expected vs actual behavior                                                                                       █
│  • Any error messages or logs                                                                                        █
│                                                                                                                      █
│  License:                                                                                                            █
│                                                                                                                      █
│  GitType is open-source software.                                                                                    ▼
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
                                                                                                                        
                 ★ Star us on GitHub (https://github.com/unhappychoice/gittype) if you enjoy GitType! ★                 
                                                                                                                        
                                [←→/HL] Switch tabs [↑↓/JK] Scroll [G] GitHub [ESC] Close
